File: packages/runtime/src/symlink.ts
import * as tg from "./index.ts";

export let symlink = async (arg: Symlink.Arg): Promise<Symlink> => {
	return await Symlink.new(arg);
};

export class Symlink {
	#state: Symlink.State;

	constructor(state: Symlink.State) {
		this.#state = state;
	}

	get state(): Symlink.State {
		return this.#state;
	}

	static withId(id: Symlink.Id): Symlink {
		return new Symlink({ id, stored: true });
	}

	static withObject(object: Symlink.Object): Symlink {
		return new Symlink({ object, stored: false });
	}

	static fromData(data: Symlink.Data): Symlink {
		return Symlink.withObject(Symlink.Object.fromData(data));
	}

	static async new(arg_: tg.Unresolved<Symlink.Arg>): Promise<Symlink> {
		let arg = await Symlink.arg(arg_);
		if ("graph" in arg) {
			return Symlink.withObject(
				arg as {
					graph: tg.Graph;
					node: number;
				},
			);
		} else {
			return Symlink.withObject({
				artifact: arg.artifact,
				path: arg.path,
			});
		}
	}

	static async arg(
		arg: tg.Unresolved<Symlink.Arg>,
	): Promise<Symlink.ArgObject> {
		let resolved = await tg.resolve(arg);
		if (typeof resolved === "string") {
			return { path: resolved };
		} else if (tg.Artifact.is(resolved)) {
			return { artifact: resolved };
		} else if (resolved instanceof tg.Template) {
			tg.assert(resolved.components.length <= 2);
			let [firstComponent, secondComponent] = resolved.components;
			if (typeof firstComponent === "string" && secondComponent === undefined) {
				return { path: firstComponent };
			} else if (
				tg.Artifact.is(firstComponent) &&
				secondComponent === undefined
			) {
				return { artifact: firstComponent, path: undefined };
			} else if (
				tg.Artifact.is(firstComponent) &&
				typeof secondComponent === "string"
			) {
				tg.assert(secondComponent.startsWith("/"));
				return {
					artifact: firstComponent,
					path: secondComponent.slice(1),
				};
			} else {
				throw new Error("invalid template");
			}
		} else if (resolved instanceof Symlink) {
			let artifact = await resolved.artifact();
			let path = await resolved.path();
			return { artifact, path };
		} else {
			return resolved;
		}
	}

	static expect(value: unknown): Symlink {
		tg.assert(value instanceof Symlink);
		return value;
	}

	static assert(value: unknown): asserts value is Symlink {
		tg.assert(value instanceof Symlink);
	}

	get id(): Symlink.Id {
		if (this.#state.id! !== undefined) {
			return this.#state.id;
		}
		let object = this.#state.object!;
		let data = Symlink.Object.toData(object);
		let id = syscall("object_id", { kind: "symlink", value: data });
		this.#state.id = id;
		return id;
	}

	async object(): Promise<Symlink.Object> {
		await this.load();
		return this.#state.object!;
	}

	async load(): Promise<tg.Symlink.Object> {
		if (this.#state.object === undefined) {
			let data = await syscall("object_get", this.#state.id!);
			tg.assert(data.kind === "symlink");
			let object = Symlink.Object.fromData(data.value);
			this.#state.object = object;
		}
		return this.#state.object!;
	}

	async store(): Promise<tg.Symlink.Id> {
		await tg.Value.store(this);
		return this.id;
	}

	async children(): Promise<Array<tg.Object>> {
		let object = await this.load();
		return tg.Symlink.Object.children(object);
	}

	async artifact(): Promise<tg.Artifact | undefined> {
		const object = await this.object();
		if (!("graph" in object)) {
			if (typeof object.artifact === "object" && "node" in object.artifact) {
				tg.assert(object.artifact.graph !== undefined, "missing graph");
				return await object.artifact.graph.get(object.artifact.node);
			}
			return object.artifact;
		} else {
			const node = (await object.graph.nodes())[object.node];
			tg.assert(node !== undefined, "invalid graph index");
			tg.assert(node.kind === "symlink", "expected a symlink");
			if (typeof node.artifact === "object" && "node" in node.artifact) {
				return await (node.artifact.graph ?? object.graph).get(
					node.artifact.node,
				);
			}
			return node.artifact;
		}
	}

	async path(): Promise<string | undefined> {
		const object = await this.object();
		if ("graph" in object) {
			const nodes = await object.graph.nodes();
			const node = nodes[object.node];
			tg.assert(node !== undefined, `invalid index ${object.node}`);
			tg.assert(
				node.kind === "symlink",
				`expected a symlink node, got ${node}`,
			);
			return node.path;
		} else {
			return object.path;
		}
	}

	async resolve(): Promise<tg.Artifact | undefined> {
		let artifact = await this.artifact();
		if (artifact instanceof Symlink) {
			artifact = await artifact.resolve();
		}
		let path = await this.path();
		if (artifact === undefined && path !== undefined) {
			throw new Error("cannot resolve a symlink with no artifact");
		} else if (artifact !== undefined && path === undefined) {
			return artifact;
		} else if (artifact instanceof tg.Directory && path !== undefined) {
			return await artifact.tryGet(path);
		} else {
			throw new Error("invalid symlink");
		}
	}
}

export namespace Symlink {
	export type Arg = string | tg.Artifact | tg.Template | Symlink | ArgObject;

	export type ArgObject =
		| { graph: tg.Graph; node: number }
		| {
				artifact?: tg.Artifact | undefined;
				path?: string | undefined;
		  };

	export type Id = string;

	export type Object =
		| { graph: tg.Graph; node: number }
		| {
				artifact: tg.Graph.Object.Edge<tg.Artifact> | undefined;
				path: string | undefined;
		  };

	export namespace Object {
		export let toData = (object: Object): Data => {
			if ("graph" in object) {
				return {
					graph: object.graph.id,
					node: object.node,
				};
			} else {
				let output: Data = {};
				if (object.artifact !== undefined) {
					output.artifact =
						"node" in object.artifact
							? { graph: object.artifact.graph?.id, node: object.artifact.node }
							: object.artifact.id;
				}
				if (object.path !== undefined) {
					output.path = object.path;
				}
				return output;
			}
		};

		export let fromData = (data: Data): Object => {
			if ("graph" in data) {
				return {
					graph: tg.Graph.withId(data.graph),
					node: data.node,
				};
			} else {
				return {
					artifact:
						typeof data.artifact === "object" && "node" in data.artifact
							? {
									graph: data.artifact.graph
										? tg.Graph.withId(data.artifact.graph)
										: undefined,
									node: data.artifact.node,
								}
							: typeof data.artifact === "string"
								? tg.Artifact.withId(data.artifact)
								: undefined,
					path: data.path,
				};
			}
		};

		export let children = (object: Object): Array<tg.Object> => {
			if ("graph" in object) {
				return [object.graph];
			} else if ("artifact" in object && object.artifact !== undefined) {
				if ("node" in object.artifact) {
					if (object.artifact.graph !== undefined) {
						return [object.artifact.graph];
					}
				} else {
					return [object.artifact];
				}
			}
			return [];
		};
	}

	export type Data =
		| { graph: tg.Graph.Id; node: number }
		| {
				artifact?: tg.Graph.Data.Edge<tg.Artifact.Id>;
				path?: string;
		  };

	export type State = tg.Object.State<Symlink.Id, Symlink.Object>;
}

File: packages/clients/python/tangram/symlink.py
"""Symlink class for Tangram.

This module mirrors the JavaScript packages/clients/js/src/symlink.ts.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Union

from tangram.object_ import Object_


async def symlink(arg: Any) -> "Symlink":
    """Create a new symlink."""
    return await Symlink.new(arg)


class Symlink:
    """A symlink is an artifact that points to another location."""

    _state: Object_.State

    def __init__(
        self,
        *,
        id: Optional[str] = None,
        object: Optional[Dict[str, Any]] = None,
        stored: bool = False,
    ) -> None:
        """Initialize a symlink."""
        obj = {"kind": "symlink", "value": object} if object is not None else None
        self._state = Object_.State(id=id, object=obj, stored=stored)

    @property
    def state(self) -> Object_.State:
        """Get the object state."""
        return self._state

    @classmethod
    def with_id(cls, id: str) -> "Symlink":
        """Create a symlink with the given ID."""
        return cls(id=id, stored=True)

    @classmethod
    def with_object(cls, object: Dict[str, Any]) -> "Symlink":
        """Create a symlink with the given object."""
        return cls(object=object, stored=False)

    @classmethod
    def from_data(cls, data: Dict[str, Any]) -> "Symlink":
        """Create a symlink from serialized data."""
        return cls.with_object(Symlink.Object.from_data(data))

    @classmethod
    async def new(cls, arg: Any) -> "Symlink":
        """Create a new symlink from an argument."""
        resolved = await cls.arg(arg)
        return cls.with_object({
            "artifact": resolved.get("artifact"),
            "path": resolved.get("path"),
        })

    @classmethod
    async def arg(cls, arg: Any) -> Dict[str, Any]:
        """Process symlink argument."""
        import tangram as tg

        resolved = await tg.resolve(arg)

        if isinstance(resolved, str):
            return {"path": resolved}
        elif _is_artifact(resolved):
            return {"artifact": resolved}
        elif isinstance(resolved, tg.Template):
            components = resolved.components
            if len(components) > 2:
                raise RuntimeError("invalid template for symlink")
            first_component = components[0] if components else None
            second_component = components[1] if len(components) > 1 else None

            if isinstance(first_component, str) and second_component is None:
                return {"path": first_component}
            elif _is_artifact(first_component) and second_component is None:
                return {"artifact": first_component, "path": None}
            elif _is_artifact(first_component) and isinstance(second_component, str):
                if not second_component.startswith("/"):
                    raise RuntimeError("path must start with /")
                return {"artifact": first_component, "path": second_component[1:]}
            else:
                raise RuntimeError("invalid template")
        elif isinstance(resolved, Symlink):
            artifact = await resolved.artifact()
            path = await resolved.path()
            return {"artifact": artifact, "path": path}
        elif isinstance(resolved, dict):
            return resolved
        else:
            raise RuntimeError(f"unsupported symlink argument type: {type(resolved)}")

    @staticmethod
    def expect(value: Any) -> "Symlink":
        """Assert that value is a Symlink and return it."""
        from tangram.assert_ import assert_

        assert_(isinstance(value, Symlink))
        return value

    @staticmethod
    def assert_(value: Any) -> None:
        """Assert that value is a Symlink."""
        from tangram.assert_ import assert_

        assert_(isinstance(value, Symlink))

    @property
    def id(self) -> str:
        """Get the symlink ID."""
        from tangram.assert_ import assert_

        id = self._state.id
        assert_(Object_.Id.kind(id) == "symlink")
        return id

    async def object(self) -> Dict[str, Any]:
        """Get the symlink object."""
        from tangram.assert_ import assert_

        obj = await self._state.load()
        assert_(obj.get("kind") == "symlink")
        return obj.get("value", obj)

    async def load(self) -> Dict[str, Any]:
        """Load the symlink object."""
        from tangram.assert_ import assert_

        obj = await self._state.load()
        assert_(obj.get("kind") == "symlink")
        return obj.get("value", obj)

    def unload(self) -> None:
        """Unload the symlink data."""
        self._state.unload()

    async def store(self) -> str:
        """Store the symlink and return its ID."""
        import tangram as tg

        await tg.Value.store(self)
        return self.id

    async def children(self) -> List[Object_]:
        """Get child objects."""
        return await self._state.children()

    async def artifact(self) -> Optional[Object_]:
        """Get the artifact this symlink points to."""
        obj = await self.object()
        artifact = obj.get("artifact")
        if artifact is None:
            return None
        if isinstance(artifact, Object_):
            return artifact
        if isinstance(artifact, str):
            return Object_.with_id(artifact)
        return None

    async def path(self) -> Optional[str]:
        """Get the path component of this symlink."""
        obj = await self.object()
        return obj.get("path")

    async def resolve(self) -> Optional[Object_]:
        """Resolve the symlink to its target artifact.

        Recursively resolves symlinks and follows paths into directories.
        """
        from tangram.directory import Directory

        artifact = await self.artifact()

        if isinstance(artifact, Symlink):
            artifact = await artifact.resolve()

        path = await self.path()

        if artifact is None and path is not None:
            raise RuntimeError("cannot resolve a symlink with no artifact")
        elif artifact is not None and path is None:
            return artifact
        elif isinstance(artifact, Directory) and path is not None:
            return await artifact.try_get(path)
        else:
            raise RuntimeError("invalid symlink")

    # Nested Object namespace.
    class Object:
        """Namespace for Symlink.Object operations."""

        @staticmethod
        def to_data(obj: Dict[str, Any]) -> Dict[str, Any]:
            """Convert object to data."""
            data: Dict[str, Any] = {}
            artifact = obj.get("artifact")
            if artifact is not None:
                if hasattr(artifact, "id"):
                    data["artifact"] = artifact.id
                else:
                    data["artifact"] = artifact
            path = obj.get("path")
            if path is not None:
                data["path"] = path
            return data

        @staticmethod
        def from_data(data: Dict[str, Any]) -> Dict[str, Any]:
            """Convert data to object."""
            obj: Dict[str, Any] = {}
            artifact_data = data.get("artifact")
            if artifact_data is not None:
                if isinstance(artifact_data, str):
                    obj["artifact"] = _artifact_with_id(artifact_data)
                else:
                    obj["artifact"] = artifact_data
            else:
                obj["artifact"] = None
            obj["path"] = data.get("path")
            return obj

        @staticmethod
        def children(obj: Optional[Dict[str, Any]]) -> List[Object_]:
            """Get child objects from symlink object."""
            if obj is None:
                return []
            artifact = obj.get("artifact")
            if artifact is not None and isinstance(artifact, Object_):
                return [artifact]
            return []


def _is_artifact(value: Any) -> bool:
    """Check if value is an Artifact."""
    from tangram.directory import Directory
    from tangram.file import File

    return isinstance(value, (Directory, File, Symlink))


def _artifact_with_id(id: str) -> Object_:
    """Create an artifact from its ID."""
    from tangram.directory import Directory
    from tangram.file import File

    prefix = id[:3] if len(id) >= 3 else ""
    if prefix == "dir":
        return Directory.with_id(id)
    elif prefix == "fil":
        return File.with_id(id)
    elif prefix == "sym":
        return Symlink.with_id(id)
    else:
        raise ValueError(f"invalid artifact id: {id}")


# Type aliases.
Symlink.Id = str
Symlink.Arg = Union[str, Object_, "tg.Template", "Symlink", Dict[str, Any]]
Symlink.Data = Dict[str, Any]
